**A method to distinguish meiotic and mitotic origins of trisomies to improve embryo selection with preimplantation genetic testing**

Aneuploidy is the leading cause of pregnancy loss and human congenital disorders, and may arise by meiotic or mitotic mechanisms. While meiotic errors affect all cells and are deleterious, mitotic errors generate mosaicism, which may be compatible with healthy live birth. Knowledge of the meiotic versus mitotic errors could therefore improve the efficacy of preimplantation genetic testing for aneuploidy (PGT-A), which seeks to rank in vitro fertilized (IVF) embryos for transfer based on their inferred ploidy status. To this end, we developed a statistical approach tailored to low-coverage sequencing-based PGT-A data to distinguish meiotic and mitotic-origin trisomies.

We compare the likelihood of the observed data under two competing hypotheses. The first assumes the presence of three unmatched homologs, an unambiguous signature of meiotic error via a heterodisomic gamete. The second assumes the presence of two identical homologs with a third unmatched homolog. This latter scenario can arise either via mitotic error or via rare meiotic errors in the absence of recombination. Our statistical models presume that the probability of two reads being drawn from the same haplotype differs under the two scenarios (1/3 and 5/9, respectively). Distinguishing these scenarios from low-coverage data (0.05 - 0.5x) is challenging since reads rarely overlap one another at SNPs. Our model thus leverages the association of alleles at different loci from large population reference panels. Hence, the trisomic chromosome is divided into Linkage disequilibrium (LD) blocks. For each LD block, several reads are chosen and the probabilities of observing them under the two models are calculated. The distribution of Log-likelihood ratios across all LD blocks is used to assess significance.

Through simulation, we find that our method retains the ability to distinguish meiotic and mitotic trisomies down to coverage as low as 0.05x. We assess the robustness of our method to admixture in the target sample, as well as mismatches between the ancestries of the target sample and the reference panel. Together, our method provides insight into the mechanistic origins of trisomies in human development, as well as a practical tool with the potential to improve embryo selection during IVF.
